### PR TITLE
PolyafceQuery methods to test for consistent dihedral angles, hence convexity

### PR DESCRIPTION
**Pull request by EarlinLutz**
New query methods for polyface:
  * PolyfaceQuery.isConvexByDihedralAngleCount (mesh, ignoreBoundaries = false) --- return true if the mesh is convex by the condition that all dihedral angles are outward.
  * PolyfaceQuery.dihedralAngleSummary(mesh, ignoreBoundaries = false) -- return plus 1 if all dihedrals are outward, -1 if all inward, 0 if mixed.

In both methods "ignoreBoundaries" means:
  * false (default) means that any boundary edge (appearing on only one facet) causes "isConvexByDihedralAngle" to be false and "dihedralAngleSummary" to be 0.
  * true means boundary (unmatched) edges are ignored in the decisions.

### CODE REVIEW

**Comment by pmconne** [comment]
@Mergifyio backport release/3.5.x

File: core/geometry/src/polyface/PolyfaceQuery.ts [hunk 1]
```diff
@@ -273,6 +274,74 @@ export class PolyfaceQuery {
...
+          edges.addEdge(visitor.clientPointIndex(i), visitor.clientPointIndex(i + 1), normalCounter);
+        }
+        normalCounter++;
+      }
+      const badClusters: SortableEdgeCluster[] = [];
+      const manifoldClusters: SortableEdgeCluster[] = [];
+
+      edges.sortAndCollectClusters(manifoldClusters, badClusters, undefined, badClusters);
+      if (badClusters.length > 0)
+        return 0;
```
Comment (dassaf4) [inline core/geometry/src/polyface/PolyfaceQuery.ts:319]: Would be useful if this function worked on non-closed meshes too. I think all you'd need is to set 2nd arg here to undefined (to ignore singleton edges). Maybe make this optional: add a "mustBeClosed" arg to `dihedralAngleSummary` that determines whether the 2nd arg is undefined, and for which `isConvexByDihedralAngleCount` could pass true.


This test be extended to test non-closed meshes by ignoring singleton edges too.

File: core/geometry/src/polyface/PolyfaceQuery.ts [hunk 2]
```diff
@@ -273,6 +274,74 @@ export class PolyfaceQuery {
...
+      const edges = new IndexedEdgeMatcher();
+      const visitor = source.createVisitor(1);
+      visitor.reset();
+      const centroidNormal: Ray3d[] = [];
+      let normalCounter = 0;
+      while (visitor.moveToNextFacet()) {
+        const numEdges = visitor.pointCount - 1;
+        const normal = PolygonOps.centroidAreaNormal(visitor.point);
+        if (normal === undefined)
+          return 0;
```
[the remark this answers is not in the document]
Comment (dassaf4) [reply core/geometry/src/polyface/PolyfaceQuery.ts:307]: (Ignoring null faces is too unwieldy.)

[the remark this answers is not in the document]
Comment (EarlinLutz) [reply core/geometry/src/polyface/PolyfaceQuery.ts:307]: I don't want to ignore nulls (slivers).  But the new ignoreBoundaries parameters allow unclosed meshes to be convex.

File: core/geometry/src/polyface/PolyfaceQuery.ts [hunk 3]
```diff
@@ -273,6 +274,74 @@ export class PolyfaceQuery {
...
+  * * Return a number summarizing the dihedral angles in the mesh.
+  *   * Return 1 if all angles are positive or planar.  The mesh is probably convex with outward normals.
+  *   * Return -1 if all angles are negative or planar.  The mesh is probably convex with inward normals.
+  *   * Return 0 if
+  *     * angles area mixed
+  *     * any edge has other than 1 incident facet or more than 2 incident facets.
+  *     * (but null edges are permitted -- These occur naturally at edges of quads at north or south pole)
+  */
+  public static dihedralAngleSummary(source: Polyface): number {
+    {
```
[the remark this answers is not in the document]
Comment (EarlinLutz) [reply core/geometry/src/polyface/PolyfaceQuery.ts:297]: done.

File: core/geometry/src/polyface/PolyfaceQuery.ts (hunk 1 above)
Comment (EarlinLutz) [reply core/geometry/src/polyface/PolyfaceQuery.ts:319]: See ignoreBoundaries option added in later commits.

**Review by dassaf4 — APPROVED** [review]

File: core/geometry/src/polyface/PolyfaceQuery.ts [hunk 4]
```diff
@@ -273,6 +274,76 @@ export class PolyfaceQuery {
...
+    const edgeVector = Vector3d.create();
+    for (const cluster of manifoldClusters) {
+      const sideA = cluster.at(0);
+      const sideB = cluster.at(1);
+      if (sideA instanceof SortableEdge
+        && sideB instanceof SortableEdge
+        && source.data.point.vectorIndexIndex(sideA.vertexIndexA, sideA.vertexIndexB, edgeVector)) {
+        const dihedralAngle = centroidNormal[sideA.facetIndex].direction.signedAngleTo(
+          centroidNormal[sideB.facetIndex].direction, edgeVector);
+        if (dihedralAngle.isAlmostZero) numPlanar++;
```
Comment (pmconne) [inline core/geometry/src/polyface/PolyfaceQuery.ts:335]: `numPlanar` is computed but has no effect on the return value - should it?

Comment (EarlinLutz) [reply core/geometry/src/polyface/PolyfaceQuery.ts:335]: I added a return case for numPositive === numNegative === 0 and numPlanar nonzero.   All the dihedral angles are zero, so its a single plane -- call it convex.

BTW, if someone (a) sends in "ignoreBoundary" and (b) has multiple connected components that are all convex, the way the two interact could be nonconvex and would not be caught.

**Pull request merged**